### PR TITLE
expression: fold const after rewrite between

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/27827

To prevent the use of wrong collation in between rewrites, we disable the constant fold, which causes the plan change.
fold the `ge` and `le` manually.

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Please add a release note, or a 'None' if it is not needed.
```

### CODE REVIEW

**Review by wjhuang2016 — APPROVED** [review]

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -1654,6 +1654,8 @@ func (er *expressionRewriter) betweenToExpression(v *ast.BetweenExpr) {
 	}
 	l.SetCharsetAndCollation(dstCharset, dstCollation)
 	r.SetCharsetAndCollation(dstCharset, dstCollation)
+	l = expression.FoldConstant(l)
+	r = expression.FoldConstant(r)
```
Comment (xuyifangreeneyes) [inline planner/core/expression_rewriter.go:1658]: Maybe we can check whether `er.tryFoldCounter > 0` to decide `tryFold` or just `fold`

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**